app/src/main/java/com/babylon/wallet/android/data/repository/state/StateRepository.kt

```diff
@@ -0,0 +1,236 @@
+import com.babylon.wallet.android.data.gateway.model.ExplicitMetadataKey
+import com.babylon.wallet.android.data.repository.cache.database.NFTEntity.Companion.asEntity
+import com.babylon.wallet.android.data.repository.cache.database.StateDao
+import com.babylon.wallet.android.data.repository.cache.database.StateDao.Companion.resourcesCacheValidity
+import com.babylon.wallet.android.data.repository.cache.database.SyncInfo
+import com.babylon.wallet.android.data.repository.cache.database.storeAccountNFTsPortfolio
+import com.babylon.wallet.android.data.repository.cache.database.updateResourceDetails
+import com.babylon.wallet.android.data.repository.toResult
+import com.babylon.wallet.android.di.coroutines.ApplicationScope
+import com.babylon.wallet.android.di.coroutines.DefaultDispatcher
+import com.babylon.wallet.android.domain.model.DApp
+import com.babylon.wallet.android.domain.model.assets.AccountWithAssets
+import com.babylon.wallet.android.domain.model.resources.Resource
+import com.babylon.wallet.android.domain.model.resources.metadata.MetadataItem.Companion.consume
+import com.babylon.wallet.android.domain.model.resources.metadata.OwnerKeyHashesMetadataItem
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.withContext
+import rdx.works.core.InstantGenerator
+import rdx.works.profile.data.model.pernetwork.Entity
+import rdx.works.profile.data.model.pernetwork.Network
+import timber.log.Timber
+import java.math.BigDecimal
+import javax.inject.Inject
+import javax.inject.Singleton
+
+interface StateRepository {
+
+    fun observeAccountsOnLedger(accounts: List<Network.Account>, isRefreshing: Boolean): Flow<List<AccountWithAssets>>
+
+    suspend fun getMoreNFTs(account: Network.Account, resource: Resource.NonFungibleResource): Result<Resource.NonFungibleResource>
+
+    fun observeResourceDetails(resourceAddress: String, accountAddress: String?): Flow<Resource>
+
+    suspend fun getNFTDetails(resourceAddress: String, localId: String): Result<Resource.NonFungibleResource.Item>
+
+    suspend fun getOwnedXRD(accounts: List<Network.Account>): Result<Map<Network.Account, BigDecimal>>
+
+    suspend fun getEntityOwnerKeys(entities: List<Entity>): Result<Map<Entity, OwnerKeyHashesMetadataItem>>
+
+    suspend fun getDAppsDetails(definitionAddresses: List<String>): Result<List<DApp>>
+
+    sealed class NFTPageError(cause: Throwable) : Exception(cause) {
+        data object NoMorePages : NFTPageError(RuntimeException("No more NFTs for this resource."))
+
+        data object VaultAddressMissing : NFTPageError(RuntimeException("No vault address to fetch NFTs"))
+
+        data object StateVersionMissing : NFTPageError(RuntimeException("State version missing for account."))
+    }
+}
+
+class StateRepositoryImpl @Inject constructor(
+    private val stateApi: StateApi,
```

Comment:
StateApi is not a singleton, and it does not have to be. It is ok to recreate an instance of this class every time you need one.

> every time a new retrofit client is created that does exactly the same job
This is not true, the http client is a singleton and is provided once. OkHttp docs suggest exactly that. The okhttp client to be created once, but the retrofit wrapper (each request) to be created each time.

Also the repository does not need to be a singleton. It is just a request/response layer. The cache does make sense to be a singleton since now we incorporate a memory cache.